crates/syntax/src/lib.rs

```diff
@@ -0,0 +1,262 @@
+    FUNCTION_LITERAL,
+    ANONYMOUS_FUNCTION,
+    LAMBDA_PARAMETER,
+    LAMBDA_PARAMETERS,
+    LAMBDA_LITERAL,
+    MULTI_LINE_STRING_EXPRESSION,
+    MULTI_LINE_STRING_CONTENT,
+    LINE_STRING_EXPRESSION,
+    LINE_STRING_CONTENT,
+    MULTI_LINE_STRING_LITERAL,
+    LINE_STRING_LITERAL,
+    STRING_LITERAL,
+    LITERAL_CONSTANT,
+    COLLECTION_LITERAL,
+    PARENTHESIZED_EXPRESSION,
+    PRIMARY_EXPRESSION,
+    VALUE_ARGUMENT,
+    VALUE_ARGUMENTS,
+    TYPE_ARGUMENTS,
+    ANNOTATED_LAMBDA,
+    CALL_SUFFIX,
+    NAVIGATION_SUFFIX,
+    INDEXING_SUFFIX,
+    ASSIGNABLE_SUFFIX,
+    PARENTHESIZED_ASSIGNABLE_EXPRESSION,
+    ASSIGNABLE_EXPRESSION,
+    PARENTHESIZED_DIRECTLY_ASSIGNABLE_EXPRESSION,
+    DIRECTLY_ASSIGNABLE_EXPRESSION,
+    POSTFIX_UNARY_SUFFIX,
+    POSTFIX_UNARY_EXPRESSION,
+    UNARY_PREFIX,
+    PREFIX_UNARY_EXPRESSION,
+    AS_EXPRESSION,
+    MULTIPLICATIVE_EXPRESSION,
+    ADDITIVE_EXPRESSION,
+    RANGE_EXPRESSION,
+    INFIX_FUNCTION_CALL,
+    ELVIS,
+    ELVIS_EXPRESSION,
+    INFIX_OPERATION,
+    GENERIC_CALL_LIKE_COMPARISON,
+    COMPARISON,
+    EQUALITY,
+    CONJUNCTION,
+    DISJUNCTION,
+    EXPRESSION,
+    SEMIS,
+    SEMI,
+    ASSIGNMENT,
+    DO_WHILE_STATEMENT,
+    WHILE_STATEMENT,
+    FOR_STATEMENT,
+    LOOP_STATEMENT,
+    BLOCK,
+    CONTROL_STRUCTURE_BODY,
+    LABEL,
+    STATEMENT,
+    STATEMENTS,
+    DEFINITELY_NON_NULLABLE_TYPE,
+    PARENTHESIZED_USER_TYPE,
+    RECEIVER_TYPE,
+    PARENTHESIZED_TYPE,
+    FUNCTION_TYPE_PARAMETERS,
+    FUNCTION_TYPE,
+    TYPE_PROJECTION_MODIFIER,
+    TYPE_PROJECTION_MODIFIERS,
+    TYPE_PROJECTION,
+    SIMPLE_USER_TYPE,
+    USER_TYPE,
+    QUEST,
+    NULLABLE_TYPE,
+    TYPE_REFERENCE,
+    TYPE,
+    ENUM_ENTRY,
+    ENUM_ENTRIES,
+    ENUM_CLASS_BODY,
+    CONSTRUCTOR_DELEGATION_CALL,
+    SECONDARY_CONSTRUCTOR,
+    OBJECT_DECLARATION,
+    PARAMETER,
+    PARAMETER_WITH_OPTIONAL_TYPE,
+    FUNCTION_VALUE_PARAMETER_WITH_OPTIONAL_TYPE,
+    PARAMETERS_WITH_OPTIONAL_TYPE,
+    SETTER,
+    GETTER,
+    PROPERTY_DELEGATE,
+    PROPERTY_DECLARATION,
+    MULTI_VARIABLE_DECLARATION,
+    VARIABLE_DECLARATION,
+    FUNCTION_BODY,
+    FUNCTION_DECLARATION,
+    FUNCTION_VALUE_PARAMETER,
+    FUNCTION_VALUE_PARAMETERS,
+    COMPANION_OBJECT,
+    ANONYMOUS_INITIALIZER,
+    CLASS_MEMBER_DECLARATION,
+    CLASS_MEMBER_DECLARATIONS,
+    TYPE_CONSTRAINT,
+    TYPE_CONSTRAINTS,
+    TYPE_PARAMETER,
+    TYPE_PARAMETERS,
+    EXPLICIT_DELEGATION,
+    ANNOTATED_DELEGATION_SPECIFIER,
+    CONSTRUCTOR_INVOCATION,
+    DELEGATION_SPECIFIER,
+    DELEGATION_SPECIFIERS,
+    CLASS_PARAMETER,
+    CLASS_PARAMETERS,
+    CLASS_BODY,
+    PRIMARY_CONSTRUCTOR,
+    CLASS_DECLARATION,
+    DECLARATION,
+    TYPE_ALIAS,
+    TOP_LEVEL_OBJECT,
+    IMPORT_ALIAS,
+    IMPORT_HEADER,
+    IMPORT_LIST,
+    PACKAGE_HEADER,
+    FILE_ANNOTATION,
+    SHEBANG_LINE,
+    SCRIPT,
+    KOTLIN_FILE,
+    ROOT,
+}
+
+pub type SyntaxNode = rowan::SyntaxNode<Lang>;
```

Comment:
strugglinh to understand why it was necessary to use rowan. Looks like it can be done without

Reply:
Basically the idea of rowan is that we get a tree that is very efficient to store in memory. Storing a whole AST in memory would consume a lot of memory.

Reply:
Our AST will now be powered by the nodes from rowan, so the AST themselves would not need to store any data

Reply:
Every AST node would wrap a SyntaxNode and whenever they need data, say an FnDef AST node needs its parameter list, it would ask the SyntaxNode is wraps for children that match ParamDef Node

Reply:
So the goal is to have a very reduced memory footprint, while still being fast

Reply:
Alright. Have you benchmarked it or its all in theory?